core/src/main/java/hudson/util/DOSToUnixPathHelper.java

```diff
@@ -0,0 +1,75 @@
+package hudson.util;
+
+import hudson.EnvVars;
+import hudson.Util;
+
+import java.io.File;
+
+import static hudson.Util.fixEmpty;
+
+class DOSToUnixPathHelper {
+    interface Helper {
+        void ok();
+        void checkExecutable(File fexe);
+        void error(String string);
+        void validate(File fexe);
+    }
+    static private boolean checkPrefix(String prefix, Helper helper) {
+        File f = new File(prefix);
+        if(f.exists()) {
+            helper.checkExecutable(f);
+            return true;
+        }
+
+        File fexe = new File(prefix+".exe");
+        if(fexe.exists()) {
+            helper.checkExecutable(fexe);
+            return true;
+        }
+        return false;
+    }
+    static void iteratePath(String exe, Helper helper) {
+        exe = fixEmpty(exe);
```

Comment:
if this is intended to be windows only it is so much cleaner just to call `where`

Reply:
It's a purely Windows path, yes. But this is a refactoring, and the code is written based on some public best practices from MS.

If someone wants to try to switch this to using where after it's refactored so there's only a single piece of code to change, they're welcome to do so, but that shouldn't be part of this PR.